### ISSUE TITLE
PhantomPinned allows conflicting implementations of trait without warning

### ISSUE THREAD

**Issue by ratmice** [rust-lang/rust#65578]
Sorry if this is a duplicate, I looked but did not find any bug about it.
PhantomPinned appears to be not a direct equivalent of the negative trait `!Unpin`,
in that implementing both PhantomPinned does not produce an error, where the same implementation of !Unpin and Unpin does.

```
#![allow(dead_code)]
#![feature(optin_builtin_traits)]
use std::marker::PhantomPinned;

struct MyFoo<F> {
      f: F,
      _pinned: PhantomPinned,
}

// Uncomment for errors on nightly.
// impl<F> !Unpin for MyFoo<F>{}
impl<F> Unpin for MyFoo<F>{}
```

**Comment by Centril** [comment]
I don't see where the bug here is. `PhantomPinned` is a type which doesn't implement `Unpin` and will cause any type which includes it to not automatically get the auto-trait implementation of `Unpin`. This is how `auto trait`s work. It's not a direct equivalent of `!Unpin` but I don't think such a claim was made.

**Comment by ratmice** [comment]
@Centril I find it extremely concerning that PhantomPinned is the only way to specify the entrance into a contract which is violable by safe code merely by implementing Unpin.  Without a way to specify that implementing Unpin should be an error.  The fact that PhantomPinned has transitive behavior on `auto trait`, but not transitive behavior on the error, means it is terribly named.

The actual behavior of PhantomPinned is not pinning, but PhantomNotAutoUnpin.
That the compiler may *suggest* adding an impl of Unpin, for safe code which will break the contract, combined with the lack of any mechanism to reason locally about the enforcement of the contract.  This all leads to very fragile code and given minimum supported compiler versions, this fragile code is likely to be with us for a while...

**Comment by Centril** [comment]
I'm going to cc @Nemo157 @withoutboats @RalfJung but I don't think I can do more than that. :)

**Comment by RalfJung** [comment]
Yeah I can see how that's confusing. OTOH, even if `Unpin` were unsafe, `PhantomPinned` would still not protect against the implicit unpinning happening in the destructor:
```rust
struct MyFoo<F> {
      f: F,
      _pinned: PhantomPinned,
}

impl<F> Drop for MyFoo<F> {
    fn drop(&mut self) {
        // *oops*, I have unpinned access to what might be a pinned instance
    }
}
```
So, making `Unpin` unsafe would give a false sense of security.

You can only implement traits for a type inside the same crate, so you *do* have this under control and can still provide safe abstractions. What is odd is that usually the module boundary is the abstraction boundary; here we actually need to make it the entire crate.

**Comment by Centril** [comment]
> You can only implement traits for a type inside the same crate, so you _do_ have this under control and can still provide safe abstractions. What is odd is that usually the module boundary is the abstraction boundary; here we actually need to make it the entire crate.

(Unless the type isn't reachable outside the current module in which case the current module again becomes the boundary?)

**Comment by RalfJung** [comment]
Well sure... but even if you newtype it the rest of the crate could still `impl Unpin for` the newtype. And if it is entirely private to the module, well, then it's not very interesting to look at as far as safe abstractions go.

**Comment by withoutboats** [comment]
This is just how all of the features involved here (auto traits mainly) work: the same would be true of using `PhantomData<*mut ()>` to make something un-`Send`.

I think the fact that is really surprising to this user is that `Unpin` is safe to implement, but as Ralf has pointed out that is the (unfortunate) consequence of the signature of `Drop`, which is why pin projecting is always unsafe.
